proposals/nnnn-safe-shared-contiguous-storage.md

```diff
@@ -0,0 +1,1227 @@
+}
+
+extension SIMD: ContiguousStorage<Self.Scalar> {
+  var storage: Span<Self.Scalar> { get }
+}
+extension KeyValuePairs: ContiguousStorage<(Self.Key, Self.Value)> {
+  var storage: Span<(Self.Key, Self.Value)> { get }
+}
+extension CollectionOfOne: ContiguousStorage<Element> {
+  var storage: Span<Element> { get }
+}
+
+extension Slice: ContiguousStorage where Base: ContiguousStorage {
+  var storage: Span<Base.Element> { get }
+}
+```
+
+In addition to the the safe types above gaining the `storage` property, the `UnsafeBufferPointer` family of types will also gain access to a `storage` property. This enables interoperability of `Span`-taking API. While a `Span` binding created from an `UnsafeBufferPointer` exists, the memory that underlies it must not be deinitialized or deallocated.
+
+```swift
+extension UnsafeBufferPointer: ContiguousStorage<Self.Element> {
+  // note: additional preconditions apply until the end of the scope
+  var storage: Span<Element> { get }
+}
+extension UnsafeMutableBufferPointer: ContiguousStorage<Self.Element> {
+  // note: additional preconditions apply until the end of the scope
+  var storage: Span<Element> { get }
+}
+extension UnsafeRawBufferPointer: ContiguousStorage<UInt8> {
+  // note: additional preconditions apply until the end of the scope
+  var storage: Span<UInt8> { get }
+}
+extension UnsafeMutableRawBufferPointer: ContiguousStorage<UInt8> {
+  // note: additional preconditions apply until the end of the scope
+  var storage: Span<UInt8> { get }
+}
+```
+
+#### Using `Span` with C functions or other unsafe code:
+
+`Span` has an unsafe hatch for use with unsafe code.
+
+```swift
+extension Span {
+  func withUnsafeBufferPointer<Result>(
+    _ body: (_ buffer: UnsafeBufferPointer<Element>) -> Result
+  ) -> Result
+}
+
+extension Span where Element: BitwiseCopyable {
+  func withUnsafeBytes<Result>(
+    _ body: (_ buffer: UnsafeRawBufferPointer) -> Result
+  ) -> Result
+}
+```
+
+
+
+### Complete `Span` API:
+
+```swift
+public struct Span<Element: ~Copyable & ~Escapable>: Copyable, ~Escapable {
+  internal var _start: UnsafePointer<Element>
+  internal var _count: Int
+}
+```
+
+##### Creating a `Span`:
+
+The initialization of a `Span` instance from an unsafe pointer is an unsafe operation. When it is initialized correctly, subsequent uses of the borrowed instance are safe. Typically these initializers will be used internally to a container's implementation of functions or computed properties that return a borrowed `Span`.
+
+```swift
+extension Span {
+
+  /// Unsafely create a `Span` over initialized memory.
+  ///
+  /// The memory in `buffer` must be owned by the instance `owner`,
+  /// meaning that as long as `owner` is alive the memory will remain valid.
+  ///
+  /// - Parameters:
+  ///   - buffer: an `UnsafeBufferPointer` to initialized elements.
+  ///   - owner: a binding whose lifetime must exceed that of
+  ///            the newly created `Span`.
+  public init<Owner>(
```

Comment:
```suggestion
  public init<Owner: ~Copyable>(
```

Should it also be marked `~Escapable`?

Reply:
All the owners have been marked `~Copyable & ~Escapable`